File: android/app/src/main/java/com/scholarlyapps/pathlingo/ui/auth/AuthViewModel.java
package com.scholarlyapps.pathlingo.ui.auth;

import android.os.Handler;
import android.os.Looper;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.scholarlyapps.pathlingo.data.ApiResult;
import com.scholarlyapps.pathlingo.data.remote.ServiceLocator;
import com.scholarlyapps.pathlingo.data.remote.dto.AppUserDto;
import com.scholarlyapps.pathlingo.data.repo.AuthRepository;

import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AuthViewModel extends ViewModel {

    private final AuthRepository repo = ServiceLocator.authRepository;
    private final MutableLiveData<AuthUiState> stateLiveData = new MutableLiveData<>(new AuthUiState());
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    public LiveData<AuthUiState> getStateLiveData() {
        return stateLiveData;
    }

    public void login(String email, String password, String deviceName) {
        launchAuth(() -> repo.login(email, password, deviceName));
    }

    public void register(String name, String email, String password, String confirm, String deviceName) {
        launchAuth(() -> repo.register(name, email, password, confirm, deviceName));
    }

    public void guestLogin(String deviceName) {
        launchAuth(() -> repo.guestLogin(deviceName));
    }

    public void loginWithGoogle(String idToken, String deviceName) {
        launchAuth(() -> repo.loginWithGoogle(idToken, deviceName));
    }

    public void sendOtp() {
        stateLiveData.setValue(new AuthUiState(true, null, null, false, false));
        executor.execute(() -> {
            AuthUiState next;
            try {
                ApiResult<String> result = repo.sendOtp();
                next = result.isSuccess()
                        ? new AuthUiState(false, null, null, false, false)
                        : new AuthUiState(false, result.getError(), null, false, false);
            } catch (Exception e) {
                next = new AuthUiState(false, e.getMessage() != null ? e.getMessage() : "Something went wrong.", null, false, false);
            }
            AuthUiState finalNext = next;
            mainHandler.post(() -> stateLiveData.setValue(finalNext));
        });
    }

    public void verifyOtp(String code) {
        stateLiveData.setValue(new AuthUiState(true, null, null, false, false));
        executor.execute(() -> {
            AuthUiState next;
            try {
                ApiResult<String> result = repo.verifyOtp(code);
                next = result.isSuccess()
                        ? new AuthUiState(false, null, null, false, true)
                        : new AuthUiState(false, result.getError(), null, false, false);
            } catch (Exception e) {
                next = new AuthUiState(false, e.getMessage() != null ? e.getMessage() : "Something went wrong.", null, false, false);
            }
            AuthUiState finalNext = next;
            mainHandler.post(() -> stateLiveData.setValue(finalNext));
        });
    }

    private void launchAuth(Callable<ApiResult<AppUserDto>> block) {
        stateLiveData.setValue(new AuthUiState(true, null, null, false, false));
        executor.execute(() -> {
            AuthUiState next;
            try {
                ApiResult<AppUserDto> result = block.call();
                next = result.isSuccess()
                        ? new AuthUiState(false, null, result.getData(), true, false)
                        : new AuthUiState(false, result.getError(), null, false, false);
            } catch (Exception e) {
                next = new AuthUiState(false, e.getMessage() != null ? e.getMessage() : "Something went wrong.", null, false, false);
            }
            AuthUiState finalNext = next;
            mainHandler.post(() -> stateLiveData.setValue(finalNext));
        });
    }

    @Override
    protected void onCleared() {
        executor.shutdown();
    }
}

File: android/app/src/main/java/com/scholarlyapps/pathlingo/ui/auth/AuthViewModel.kt
package com.scholarlyapps.pathlingo.ui.auth

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.scholarlyapps.pathlingo.data.remote.ServiceLocator
import com.scholarlyapps.pathlingo.data.remote.dto.AppUserDto
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class AuthUiState(
    val loading: Boolean = false,
    val error: String? = null,
    val user: AppUserDto? = null,
    val success: Boolean = false,
)

class AuthViewModel : ViewModel() {

    private val repo = ServiceLocator.authRepository

    private val _state = MutableStateFlow(AuthUiState())
    val state: StateFlow<AuthUiState> = _state.asStateFlow()

    private val _stateLiveData = MutableLiveData(AuthUiState())
    val stateLiveData: LiveData<AuthUiState> = _stateLiveData

    fun login(email: String, password: String, deviceName: String) {
        launchAuth { repo.login(email, password, deviceName) }
    }

    fun register(name: String, email: String, password: String, confirm: String, deviceName: String) {
        launchAuth { repo.register(name, email, password, confirm, deviceName) }
    }

    fun guestLogin(deviceName: String) {
        launchAuth { repo.guestLogin(deviceName) }
    }

    fun loginWithGoogle(idToken: String, deviceName: String) {
        launchAuth { repo.loginWithGoogle(idToken, deviceName) }
    }

    fun consumeError() {
        _state.value = _state.value.copy(error = null)
    }

    private fun launchAuth(block: suspend () -> Result<AppUserDto>) {
        val loading = AuthUiState(loading = true)
        _state.value = loading
        _stateLiveData.value = loading
        viewModelScope.launch {
            val next = block().fold(
                onSuccess = { AuthUiState(user = it, success = true) },
                onFailure = { AuthUiState(error = it.message ?: "Something went wrong.") },
            )
            _state.value = next
            _stateLiveData.postValue(next)
        }
    }
}
